### ISSUE TITLE
Inconsistency in accuracy between different versions

### ISSUE THREAD

**Issue by jinamshah** [pytorch/fairseq#1589]
### To Reproduce

Steps to reproduce the behavior (**always include the command you ran**):

1. Run cmd 'python3.6 /path/to/fairseq/train.py --save-dir ./output/ --arch fconv --clip-norm 0.x --lr x --min-lr x --momentum 0.x --max-epoch 100 --no-progress-bar --seed 1 /path/to/input_files --encoder-embed-dim x --decoder-embed-dim x  --decoder-out-embed-dim x --dropout x --encoder-layers "specified arch" --decoder-layers "specified arch" --batch-size x --skip-invalid-size-inputs-valid-test '
2. See error

There is no error as such, as the training job runs but the model is unable to converge

### Expected behavior

The accuracy numbers should not change if the setup remains the same and only the versions of fairseq and pytorch is changed.

### Environment

 - fairseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0): 1.3
 - OS (e.g., Linux): ubuntu 18.04
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): `pip install --editable .`
 - Python version: 3.6
 - CUDA/cuDNN version: 10.1/7.6.4
 - GPU models and configuration: 8 tesla k80 on single machine

### Additional context

I have implemented a translation task using an old version of fairseq and received an accuracy of 'x'. I then upgraded the pytorch and fairseq versions to the latest versions (1.3 and master branch respectively), and then tried rerunning my training job, keeping the parameters the same. The accuracy that I received from this experiment is 'x/3'. Can someone please help me understand what has changed in particular with the translation task?

Also please note: 
the setup remains the same (multi-gpu)
the model has stopped converging like before

**Comment by myleott** [comment]
We ensure exact reproducibility when rerunning then same command using a single version of fairseq and pytorch. However, different versions of pytorch may produce slightly different results from each other, due to subtle changes in the underlying implementations of various ops. Similarly, occasionally different versions of fairseq will cause the random number generator to shift, which may cause things like dropout masks to be different. Usually you can obtain identical results if you turn off dropout, but depending on how big the jump is in versions, it may not be possible to get identical results.

**Comment by myleott** [comment]
What was the old version of fairseq and pytorch that you were using before? Also, you were using the fconv architecture?

**Comment by jinamshah** [comment]
Hey @myleott ! 
I don't have the exact version number for the earlier version I used, but I do have the commit ID, the problem is that it doesn't exist in your master branch anymore. The version is ideally 3-5 months old
Yes, I was using fconv architecture

> Usually you can obtain identical results if you turn off dropout

I will try this and update here on the same!

**Comment by myleott** [comment]
What was the commit ID?

**Comment by jinamshah** [comment]
commit ID: 7a31fe068c8015e4ee9f8d45887efcd722b53fc0

**Comment by jinamshah** [comment]
Update: Changing the fairseq versions while keeping the parameters same (without dropout) still resulted in a significant drop in the accuracy numbers.

**Comment by myleott** [comment]
Hmm, there shouldn't been any changes to the fconv implementation since 7a31fe068c8015e4ee9f8d45887efcd722b53fc0 that would have caused a difference.

A few more follow-up questions:
1) are you using the same version of pytorch for your latest comparison? The only difference is the fairseq commit?
2) one recent change is that we set a global numpy seed (4a7cd5825717b0b0c96a6463e0dd2d7b18dd4331). To test if this might be breaking things, can you try going back to your old version of fairseq and modifying train.py to set the global seed, similar to this: https://github.com/pytorch/fairseq/commit/4a7cd5825717b0b0c96a6463e0dd2d7b18dd4331#diff-e44f4a60e89e820dde9bb27afa634965. This will confirm whether it's just randomness that is causing the discrepancy or something else.

Also, can you share the training log before and after? Is training converting more slowly, or does the loss explode?

**Comment by jinamshah** [comment]
Hi @myleott ,
1. pytorch version for the previous setup was 0.4.0, and for the new version it is 1.3
2. there was no difference in the accuracy numbers from using the suggestions provided by you in point 2 (using numpy random seed in the old version of fairseq, and I also tried removing the numpy seed from the new version. None of these had any effect)
3. I have noted that training is a lot faster for the new version and there is a loss drops drastically after the first epoch

A question: will the preprocessing step affect the training job significantly? i.e. the difference in the preprocessing(preprocess.py file) between the two versions
I have a feeling that there is a problem in preprocessing step here

**Comment by myleott** [comment]
1. You were using pytorch 0.4.0 with commit 7a31fe0? I'm surprised that fairseq from 4 months ago would work with such an old version of pytorch... please confirm.
2. 👍 
3. 👍 
4. While we've made changes to improve the speed of preprocess.py, the semantics are unchanged. We still use datasets preprocessed 1+ year ago and everything works.

Are you able to share your training logs?

**Comment by jinamshah** [comment]
Hi @myleott 
1. yes
4. the numbers change drastically on changing the preprocessing step. If I use preprocessing from the previous version mentioned, accuracy is 'x' and if I use the new version, the accuracy is 'x/3'

**Comment by myleott** [comment]
What is your preprocessing command? It's difficult to debug without more details about the dataset/preprocessing steps.

**Comment by jinamshah** [comment]
@myleott this the command I used for both the versions of fairseq
`python3.6 ./preprocess.py --source-lang src --target-lang trg --trainpref data_dir/train --validpref data_dir/dev --testpref  data_dir/dev --nwordssrc x(int) --nwordstgt x(int) --destdir data_dir/bin  --srcdict '/path/to/src_dict' --tgtdict '/path/to/trg_dict'`

**Comment by jinamshah** [comment]
@myleott can you please tell me what changes have been made specific to fconv architecture?

**Comment by myleott** [comment]
There haven't been any changes to fconv in a long time: https://github.com/pytorch/fairseq/commits/master/fairseq/models/fconv.py

Nor have there been changes to the ConvTBD or LinearizedConvolution layers: https://github.com/pytorch/fairseq/commits/master/fairseq/modules/conv_tbc.py and https://github.com/pytorch/fairseq/commits/master/fairseq/modules/linearized_convolution.py

Without a training log, I'm unsure how to help you debug this further.
